README.md
# flutter_acpplaces

[![pub package](https://img.shields.io/pub/v/flutter_acpplaces.svg)](https://pub.dartlang.org/packages/flutter_acpplaces) ![Build](https://github.com/adobe/flutter-acpplaces/workflows/Dart%20Unit%20Tests%20+%20Android%20Build%20+%20iOS%20Build/badge.svg) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

`flutter_acpplaces` is a flutter plugin for the iOS and Android [AEP Places SDK](https://docs.adobe.com/content/help/en/places/using/places-ext-aep-sdks/places-extension/places-extension.html) to allow for integration with flutter applications. Functionality to enable the Places extension is provided entirely through Dart documented below.

## Installation

Install instructions for this package can be found [here](https://pub.dev/packages/flutter_acpplaces#-installing-tab-).

> Note: After you have installed the SDK, don't forget to run `pod install` in your `ios` directory to link the libraries to your Xcode project.

## Tests

Run:

```bash
flutter test
```

## Usage

### [Places](https://docs.adobe.com/content/help/en/places/using/places-ext-aep-sdks/places-extension/places-extension.html)

##### Importing the SDK:
```dart
import 'package:flutter_acpplaces/flutter_acpplaces.dart';
```

##### Getting the SDK version:
 ```dart
String version = await FlutterACPPlaces.extensionVersion;
 ```

 ##### Registering the extension with ACPCore:

 > Note: It is required to initialize the SDK via native code inside your AppDelegate and MainApplication for iOS and Android respectively. For more information see how to initialize [Core](https://aep-sdks.gitbook.io/docs/getting-started/get-the-sdk#2-add-initialization-code).

 ##### **iOS**
Swift
 ```swift
import ACPPlaces

ACPPlaces.registerExtension()
 ```
Objective-C
 ```objective-c
#import "ACPPlaces.h"

[ACPPlaces registerExtension];
 ```

 ##### **Android:**
 ```java
import com.adobe.marketing.mobile.Places;

Places.registerExtension();
 ```

##### Clear client side Places plugin data:

```dart
try {
  await FlutterACPPlaces.clear;
} on PlatformException {
  log("Failed to clear Places data.");
}
```

##### Get the current POI's that the device is currently known to be within:

```dart
String pois;
try {
  pois = await FlutterACPPlaces.currentPointsOfInterest;
} on PlatformException {
  log("Failed to get the current POI's.");
}
```

##### Get the last latitude and longitude stored in the Places plugin:

```dart
String location;
try {
  location = await FlutterACPPlaces.lastKnownLocation;
} on PlatformException {
  log("Failed to get the last known location.");
}
```

##### Get a list of nearby POI's:

```dart
String pois;
try {
  var location = {'latitude':37.3309422, 'longitude': -121.8939077};
  pois = await FlutterACPPlaces.getNearbyPointsOfInterest(location, 100);
} on PlatformException {
  log("Failed to get the nearby POI's.");
}
```

##### Pass a Geofence and transition type to be processed by the Places plugin:

```dart
var geofence = new Geofence({'requestId':'d4e72ade-0400-4280-9bfe-8ba7553a6444', 'latitude':37.3309422, 'longitude': -121.8939077, 'radius': 1000, 'expirationDuration':-1});
FlutterACPPlaces.processGeofence(geofence, ACPPlacesRegionEventType.ENTRY);
```

##### Set the authorization status:

```dart
FlutterACPPlaces.setAuthorizationStatus(ACPPlacesAuthorizationStatus.ALWAYS);
```

## Contributing

See [CONTRIBUTING](CONTRIBUTING.md)

## License
See [LICENSE](LICENSE)


CHANGELOG.md
## 2.0.1
* Migrate Android library from jcenter to mavenCentral.

## 2.0.0
* Null-safety support

# 1.0.1
* Update the package description and bump version to attempt to resolve "No valid SDK" error seen at pub.dev.

# 1.0.0
* Initial release of Adobe Experience Platform - Places plugin for Flutter apps


example/lib/main.dart
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import 'dart:async';
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_acpplaces/flutter_acpplaces.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';
  String _currentpois = "null";
  String _nearbypois = "null";
  String _location = "null";

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      platformVersion = await FlutterACPPlaces.extensionVersion;
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  Future<void> clear() async {
    try {
      await FlutterACPPlaces.clear();
    } on PlatformException {
      log("Failed to clear Places data.");
    }
  }

  Future<void> getCurrentPointsOfInterest() async {
    String pois = '';
    try {
      pois = await FlutterACPPlaces.currentPointsOfInterest ?? '???';
    } on PlatformException {
      log("Failed to get the current POI's.");
    }

    if (!mounted) return;
    setState(() {
      _currentpois = pois.toString();
    });
  }

  Future<void> getLastKnownLocation() async {
    String location = '';
    try {
      location = await FlutterACPPlaces.lastKnownLocation ?? '???';
    } on PlatformException {
      log("Failed to get the last known location.");
    }

    if (!mounted) return;
    setState(() {
      _location = location;
    });
  }

  Future<void> getNearbyPointsOfInterest() async {
    String pois = '';
    try {
      var location = {'latitude': 37.3309422, 'longitude': -121.8939077};
      pois = await FlutterACPPlaces.getNearbyPointsOfInterest(location, 100) ??
          '???';
    } on PlatformException {
      log("Failed to get the nearby POI's.");
    }

    if (!mounted) return;
    setState(() {
      _nearbypois = pois.toString();
    });
  }

  Future<void> processGeofence() async {
    var geofence = new Geofence({
      'requestId': 'd4e72ade-0400-4280-9bfe-8ba7553a6444',
      'latitude': 37.3309422,
      'longitude': -121.8939077,
      'radius': 1000,
      'expirationDuration': -1
    });
    FlutterACPPlaces.processGeofence(geofence, ACPPlacesRegionEventType.entry);
  }

  Future<void> setAuthorizationStatus() async {
    FlutterACPPlaces.setAuthorizationStatus(
        ACPPlacesAuthorizationStatus.always);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        home: Scaffold(
      appBar: AppBar(
        title: const Text('ACPPlaces Plugin example app'),
      ),
      body: Center(
        child: Column(children: <Widget>[
          Text('ACPPlaces version = $_platformVersion\n'),
          Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                ElevatedButton(
                  child: Text("ACPPlaces.getCurrentPointsOfInterest()"),
                  onPressed: () => getCurrentPointsOfInterest(),
                ),
              ]),
          Flexible(
            child: new Text('Current Points of Interest = $_currentpois\n'),
          ),
          Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                ElevatedButton(
                  child: Text("ACPPlaces.getNearbyPointsOfInterest()"),
                  onPressed: () => getNearbyPointsOfInterest(),
                ),
              ]),
          Flexible(
            child: new Text('Nearby Points of Interest = $_nearbypois\n'),
          ),
          ElevatedButton(
            child: Text("ACPPlaces.clear()"),
            onPressed: () => FlutterACPPlaces.clear(),
          ),
          ElevatedButton(
            child: Text("ACPPlaces.setAuthorizationStatus()"),
            onPressed: () => setAuthorizationStatus(),
          ),
          Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                ElevatedButton(
                  child: Text('ACPPlaces.getLastKnownLocation()'),
                  onPressed: () => getLastKnownLocation(),
                ),
              ]),
          Flexible(
            child: new Text('Last known location = $_location\n'),
          ),
          Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                ElevatedButton(
                  child: Text("ACPPlaces.processGeofence()"),
                  onPressed: () => processGeofence(),
                ),
              ]),
        ]),
      ),
    ));
  }
}
